hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/Planner.java

```diff
@@ -55,13 +62,69 @@
     DAG createDag() {
         Map<Transform, List<Transform>> adjacencyMap = pipeline.adjacencyMap();
         validateNoLeakage(adjacencyMap);
+
+        // Calculate greatest common denominator of frame lengths from all transforms in the pipeline
+        long[] gcdFrame = calculateGcd(adjacencyMap.keySet().stream()
+                                                .map(Transform::watermarkFrameDef)
+                                                .filter(frame -> frame[0] > 0)
+                                                .collect(toList()));
+        WatermarkEmissionPolicy emitPolicy = gcdFrame[0] > 0
+                ? emitByFrame(tumblingWinPolicy(gcdFrame[0]).withOffset(gcdFrame[1]))
+                : noThrottling();
+        // Replace emission policy
+        for (Transform transform : adjacencyMap.keySet()) {
+            if (transform instanceof StreamSourceTransform) {
+                StreamSourceTransform t = (StreamSourceTransform) transform;
+                if (t.getWmParams() != null) {
+                    t.setWmGenerationParams(t.getWmParams().withEmitPolicy(emitPolicy));
+                }
+            } else if (transform instanceof TimestampTransform) {
+                TimestampTransform t = (TimestampTransform) transform;
+                t.setWmGenerationParams(t.getWmGenParams().withEmitPolicy(emitPolicy));
+            }
+        }
+
         Iterable<Transform> sorted = topologicalSort(adjacencyMap, Object::toString);
         for (Transform transform : sorted) {
             transform.addToDag(this);
         }
         return dag;
     }
 
+    // package-visible for tests
+    static long[] calculateGcd(List<long[]> frameDefs) {
```

Comment:
this could be encapsulated in another class I think, as it's not specific to planner